design/permissions.md

```diff
@@ -0,0 +1,181 @@
+	If permission decisions are persistent, load them into memory now. Default provided implementation does not persist.
+3. [Host] The host wishes to run an MRE, so creates an IMixedRealityExtensionApp instance and calls Startup().
+4. [Runtime] Downloads the manifest from the provided MRE server. If missing, assumes no permissions required.
+5. [Runtime] Calls into the permission manager requesting the manifest-listed permissions.
+6. [Runtime|Host] Permission manager evaluates the required and optional MRE permissions against the current set of
+	grants/denials for the MRE's origin, and determines if any new permission decisions need to be made by the user.
+	If so:
+	1. [Host] Present the choices to the user by whatever means the host sees fit, persist the decision if desired,
+		and report the result to the Runtime. Optional permissions should be presented in such a way as they can be
+		granted or denied individually, but required permissions should be decided as a group.
+7. [Runtime] Determine if the MRE has sufficient permission to run, i.e. `execution` and all the required permissions
+	are granted. If not, abort connecting to the MRE.
+8. [Runtime] Startup proceeds like normal, but if a `user-joined` message is sent, the user payload must include
+	the IDs of any permissions that were granted to this MRE by this user (other than `execution`, which is implied).
+
+
+Permissions Manager
+---------------------
+
+Host apps must provide hooks for the MRE subsystem to obtain permission from users. This is provided to `MREAPI.InitializeAPI`
+as an implementation of `IPermissionManager`, which has the following methods:
+
+* `Task<Permissions> PromptForPermission(...)` - Request permissions from the user, and return a Task that resolves
+	with those permissions the user has granted. Takes the following arguments:
+	* `string appOrigin` - The origin of the MRE requesting permission, as described above.
+	* `string appDisplayName` - A human-readable identifier for the MRE server provided from the manifest.
+	* `Permissions permissionsNeeded` - A bitfield of the permissions required to run the app.
+	* `Permissions permissionsWanted` - A bitfield of permissions that the app can use, but are not needed to run.
+* `Permissions CurrentPermissions(...)` - Returns a bitfield of the currently granted permissions. Takes the following
+	arguments:
+	* `string appOrigin` - The origin of the MRE for which we want the permission set.
+
+```cs
+[Flags]
+enum Permissions {
+	None = 0,
+	Execution = 1,
+	UserTracking = 2,
+	UserInteraction = 4,
+	...
+}
+```
+
+The default implementation of the permissions manager will also include the following virtual methods:
+
+* `void ReadFromStorage()` - Populate the in-memory permission database from offline storage.
+* `void WriteToStorage()` - Write the "remembered" portions of the in-memory permission database to offline storage.
```

Comment:
Is this storage we would be providing and interfacing with in our runtime code, since this refers to a default permission manager?  Does this mean that no permission manager is required to be supplied by the host app?

Reply:
The interface does not require any kind of persistent storage. But since most hosts will probably want it, the MRE namespace will also have an abstract base implementation that provides some stubs. Haven't worked out all the details here yet.

---
In reply to: [441648519](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441648519) [](ancestors = 441648519)

Reply:
Ok.  So the stubs are not for a default implementation of local storage then.  They are just for optional hooks to the permission manager supplied by the host app?

Reply:
Correct.

---
In reply to: [441735090](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441735090) [](ancestors = 441735090)